### PR TITLE
[DNM] MSBuild Task for SDL Automation

### PR DESCRIPTION
**Pull request by sunandabalu**
Initial Task for SDL tool Install
Basic Wiring into Arcade Build Definition.

### CODE REVIEW

File: src/Microsoft.DotNet.SDLAutomator/sdk/Sdk.props [hunk 1]
```diff
@@ -0,0 +1,5 @@
+<!-- Licensed to the .NET Foundation under one or more agreements. The .NET Foundation licenses this file to you under the MIT license. See the LICENSE file in the project root for more information. -->
```
Comment (tmat) [inline src/Microsoft.DotNet.SDLAutomator/sdk/Sdk.props]: What's the benefit of this being an SDK package?

File: eng/common/templates/job/job.yml [hunk 2]
```diff
@@ -195,3 +198,29 @@ jobs:
...
+        inputs:
+          command: restore
+          feedsToUse: config
+          restoreSolution: 'eng\common\internal\Tools.csproj'
+          nugetConfigPath: 'NuGet.config'
+          restoreDirectory: '$(Build.SourcesDirectory)\.packages'
+          env:
+            VSS_NUGET_EXTERNAL_FEED_ENDPOINTS: '{"endpointCredentials": [{"endpoint":"https://securitytools.pkgs.visualstudio.com/_packaging/Guardian/nuget/v3/index.json", "username":"dn-bot","password":"$(dn-bot-devdiv-nuget-feed-read)"}]}'
+            TargetGroup: $(_framework)
+      - task: PowerShell@2
```
Comment (tmat) [inline eng/common/templates/job/job.yml]: Why is this needed? Isn't it already restored via `eng\common\internal\Tools.csproj`?

File: src/Microsoft.DotNet.Arcade.Sdk/tools/SdkTasks/SDLAutomation.proj [hunk 3]
```diff
@@ -0,0 +1,30 @@
+<!-- Licensed to the .NET Foundation under one or more agreements. The .NET Foundation licenses this file to you under the MIT license. See the LICENSE file in the project root for more information. -->
```
Comment (tmat) [inline src/Microsoft.DotNet.Arcade.Sdk/tools/SdkTasks/SDLAutomation.proj]: Why is this and sdk task?

File: src/Microsoft.DotNet.SDLAutomator/InstallGuardian.cs [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+                Log.LogErrorFromException(e, true);
+            }
+            return !Log.HasLoggedErrors;
+        }
+
+        public void AddGuardianToPath()
+        {
+            try
+            {
+                const string name = "PATH";
```
Comment (tmat) [inline src/Microsoft.DotNet.SDLAutomator/InstallGuardian.cs]: This is bad. Don't set environment variables in msbuild task.

Comment (jonfortescue) [reply src/Microsoft.DotNet.SDLAutomator/InstallGuardian.cs]: A way to do this safely (for AzDO only) would be to log `##vso[task.prependpath]$(PackagesDirectory)` to the console. Not sure if that variable reference will work per se but something like that should work.

**Comment by sunandabalu** [comment]
> What is the purpose of this task?

The purpose of this task is to provide ability to run SDL tools using an internal tool called Guardian. This task is restoring the Guardian package from a secure nuget feed and adding Guardian to PATH. There are more tasks to be added which would run guardian commands to download and execute SDL tools.

File: eng/common/templates/job/job.yml (hunk 2 above)
Comment (sunandabalu) [reply eng/common/templates/job/job.yml]: Tools.csproj is restoring the Guardian package, this is to restore SDL automator pacakge and execute the tasks underneath it. Currently the SDL automator only has the install tasks, we are adding the init, run and process tasks subsequently.

File: src/Microsoft.DotNet.SDLAutomator/sdk/Sdk.props (hunk 1 above)
Comment (sunandabalu) [reply src/Microsoft.DotNet.SDLAutomator/sdk/Sdk.props]: This package needs to get called from build(job.yml) as well as release definitions.

File: src/Microsoft.DotNet.Arcade.Sdk/tools/SdkTasks/SDLAutomation.proj [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    <_MicrosoftDotNetSDLAutomatorTaskDir Condition="'$(MSBuildRuntimeType)' == 'Core'">$(_MicrosoftDotNetSDLAutomatorTaskBaseDir)netcoreapp2.1/</_MicrosoftDotNetSDLAutomatorTaskDir>
+  </PropertyGroup>
+
+  <UsingTask TaskName="InstallGuardian" AssemblyFile="$(_MicrosoftDotNetSDLAutomatorTaskDir)Microsoft.DotNet.SDLAutomator.dll"/>
+
+  <Target Name="Execute">
+    <Error Text="The PackageDirectory property wasn't supplied."
+          Condition="'$(PackageDirectory)' == ''" />
+
+    <InstallGuardian PackageDirectory="$(PackageDirectory)"
```
Comment (jonfortescue) [inline src/Microsoft.DotNet.Arcade.Sdk/tools/SdkTasks/SDLAutomation.proj]: I fix this in my PR to the branch, but the parameter in your task is `PackagesDirectory` (with an `s`)

Comment (tmat) [reply src/Microsoft.DotNet.Arcade.Sdk/tools/SdkTasks/SDLAutomation.proj]: Does it need to be added to PATH? I would rather avoid that.

File: src/Microsoft.DotNet.SDLAutomator/InstallGuardian.cs (hunk 4 above)
Comment (tmat) [reply src/Microsoft.DotNet.SDLAutomator/InstallGuardian.cs]: We should avoid modifying the PATH.

File: eng/common/templates/job/job.yml (hunk 2 above)
Comment (tmat) [reply eng/common/templates/job/job.yml]: I don't think this is the right design.

File: src/Microsoft.DotNet.SDLAutomator/InstallGuardian.cs (hunk 4 above)
Comment (sunandabalu) [reply src/Microsoft.DotNet.SDLAutomator/InstallGuardian.cs]: The PATH is altered here only for the session, is there a better way of doing this that in msbuild task? Similar to what nuget restore task from AzDO does, restore and add to PATH.

File: src/Microsoft.DotNet.SDLAutomator/sdk/Sdk.props (hunk 1 above)
Comment (tmat) [reply src/Microsoft.DotNet.SDLAutomator/sdk/Sdk.props]: Why does it need to be called from yaml? Why can't it be executed from cibuild.cmd?

File: src/Microsoft.DotNet.SDLAutomator/InstallGuardian.cs (hunk 4 above)
Comment (jonfortescue) [reply src/Microsoft.DotNet.SDLAutomator/InstallGuardian.cs]: Apparently this method only modifies the path for the session, not for the lifetime of the machine. That should be fine as it doesn't change machine state.

File: src/Microsoft.DotNet.SDLAutomator/sdk/Sdk.props (hunk 1 above)
Comment (tmat) [reply src/Microsoft.DotNet.SDLAutomator/sdk/Sdk.props]: BTW, adding "sdk/Sdk.props" has nothing to do with ability to execute from yaml.

File: src/Microsoft.DotNet.Arcade.Sdk/tools/SdkTasks/SDLAutomation.proj (hunk 3 above)
Comment (sunandabalu) [reply src/Microsoft.DotNet.Arcade.Sdk/tools/SdkTasks/SDLAutomation.proj]: This task is called in job.yml, this was the conclusion based on my discussion with @chcosta . Any downsides for this being an sdk task?

File: src/Microsoft.DotNet.SDLAutomator/sdk/Sdk.props (hunk 1 above)
Comment (tmat) [reply src/Microsoft.DotNet.SDLAutomator/sdk/Sdk.props]: Maybe there is a confusion with how we named `SdkTasks` under Arcade. That directory should rather be called "PipelineTargets" to not cause confusion with msbuild SDK packages.

File: src/Microsoft.DotNet.Arcade.Sdk/tools/SdkTasks/SDLAutomation.proj (hunk 3 above)
Comment (chcosta) [reply src/Microsoft.DotNet.Arcade.Sdk/tools/SdkTasks/SDLAutomation.proj]: I discussed some possible options including what SDK tasks were.  I didn't specifically offer an opinion on the design (or didn't intend to).  Here's a [link](https://github.com/dotnet/core-eng/blob/16f1b1f8a6f00c722e72519573084f586e7d297c/Documentation/Security/GuardianIntegrationToArcade.md) to the doc you wrote that I found.  @sunandabalu is that the right link?  I thought there was a section about the three scenarios where Guardian would run, but I don't see in that doc.

I think @tmat's input will help solidify a clean implementation.

Comment (sunandabalu) [reply src/Microsoft.DotNet.Arcade.Sdk/tools/SdkTasks/SDLAutomation.proj]: Thanks @chcosta , that link is the how-to's about Guardian. This one is the link with the [implementation detail ](https://github.com/dotnet/core-eng/blob/16f1b1f8a6f00c722e72519573084f586e7d297c/Documentation/Security/ArcadeSecurityControlFlowDocumentation.md )

File: src/Microsoft.DotNet.Arcade.Sdk/tools/SdkTasks/SDLAutomation.proj (hunk 5 above)
Comment (sunandabalu) [reply src/Microsoft.DotNet.Arcade.Sdk/tools/SdkTasks/SDLAutomation.proj]: The rest of the tasks entail running guardian commands which require guardian be added to PATH, this is the recommended usage from the guardian team.

Comment (sunandabalu) [reply src/Microsoft.DotNet.Arcade.Sdk/tools/SdkTasks/SDLAutomation.proj]: Removed the code that sets PATH, instead we can just invoke using full path.

File: src/Microsoft.DotNet.SDLAutomator/InstallGuardian.cs (hunk 4 above)
Comment (sunandabalu) [reply src/Microsoft.DotNet.SDLAutomator/InstallGuardian.cs]: Removed code that modifies PATH

File: src/Microsoft.DotNet.SDLAutomator/sdk/Sdk.props (hunk 1 above)
Comment (sunandabalu) [reply src/Microsoft.DotNet.SDLAutomator/sdk/Sdk.props]: Changed to be not an SDK package

**Comment by sunandabalu** [comment]
Having a script to run SDL commands which can then be plugged into Arcade Release Definition is a cleaner and simpler design. So moving away from the MSBuildTask design that was previously discussed, hence closing this PR.
